include/aspect/material_model/interface.h

```diff
@@ -365,6 +398,11 @@ namespace aspect
        * in order to compute the reaction increment.
        */
       std::vector<std::vector<double> > reaction_terms;
+      
+      /**
+       * The derivatives used for forming the Jacobian, needed for the Newton solver.
+       */
+      MaterialModelDerivatives<dim> derivatives;
```

Comment:
We were thinking about a collection of shared pointers to objects derived from a common base class (or at least that was Timo's idea) and we already started to work on something like that. So we should probably all sit together and talk at some point.